### PR TITLE
Feature 55 reusable material uploading

### LINKED ISSUE #55: Reusable materials images
As a user I need to be able to add images to reusable materials. I also need to be able to view the images by clickin "Show images" button from the list and see them in a popup screen.

### PR DESCRIPTION
**Pull request by Kaemppi**
Resolves #55

### CODE REVIEW

File: src/components/survey/reusables.tsx [hunk 1]
```diff
@@ -89,6 +98,15 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
     setSelectedReusableIds([]);
   }, []);
 
+  /**
+   * On new reusable file delete
+   */
+  const onNewReusableFileDelete = (index: number) => {
```
Comment (lauriLukkarinen) [inline src/components/survey/reusables.tsx:104]:
```suggestion
  /**
   * On new reusable file delete
   *
   * @param index index
   */
  const onNewReusableFileDelete = (index: number) => {
```

File: src/components/survey/reusables.tsx [hunk 2]
```diff
@@ -98,9 +116,20 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+      const imageUrlPromises = newReusableFiles.map(async newReusableFile => {
+        const uploadData = await FileUploadUtils.upload(keycloak.token!!, newReusableFile);
+        const { xhrRequest, uploadUrl, formData, key } = uploadData;
+
+        xhrRequest.open("POST", uploadUrl, true);
+        xhrRequest.send(formData);
+        return `${uploadUrl}/${key}`;
+      });
+
+      const imagesUrls = await Promise.all(imageUrlPromises);
```
Comment (lauriLukkarinen) [inline src/components/survey/reusables.tsx:128]:
```suggestion
      const imageUrls = await Promise.all(imageUrlPromises);
```

File: src/components/survey/reusables.tsx [hunk 3]
```diff
@@ -98,9 +116,20 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+
       const createdReusable = await Api.getSurveyReusablesApi(keycloak.token).createSurveyReusable({
         surveyId: surveyId,
-        reusable: newMaterial
+        reusable: { ...newMaterial, images: imagesUrls }
```
Comment (lauriLukkarinen) [inline src/components/survey/reusables.tsx:132]:
```suggestion
        reusable: { ...newMaterial, images: imageUrls }
```

File: src/components/survey/reusables.tsx [hunk 4]
```diff
@@ -142,6 +172,68 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+   */
+  const newReusableFilesUpload = (acceptedFiles: File[]) => {
+    const updatedNewReusableFiles = FileUploadUtils.normalizeFileNames([ ...newReusableFiles, ...acceptedFiles ].splice(0, 4));
+    setNewReusableFiles(updatedNewReusableFiles);
+  };
+
+  /**
+   * Reusable file upload progress
+   * 
+   * @param acceptedFiles accepted files
```
Comment (lauriLukkarinen) [inline src/components/survey/reusables.tsx:188]:
```suggestion
   * @param updatedUploadedFile updated uploaded file
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/generic/generic-dialog.tsx [hunk 5]
```diff
@@ -10,7 +10,7 @@ interface Props {
   positiveButtonText?: string;
   cancelButtonText?: string;
   onClose: () => void;
-  onCancel: () => void;
+  onCancel?: () => void;
```
Comment (jarinykanen) [inline src/components/generic/generic-dialog.tsx:13]: Why has this been changed?

File: src/components/survey/reusables.tsx [hunk 6]
```diff
@@ -142,6 +172,68 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+      await Promise.all(fileUploadPromises);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.failToUpload, error);
+    }
+
+    onMaterialRowChange(updatedReusable);
+  };
+
+  const { getRootProps, getInputProps, open } = useDropzone({
+    // Disable click and keydown behavior
```
Comment (jarinykanen) [inline src/components/survey/reusables.tsx:229]:
```suggestion
```

File: src/components/survey/reusables.tsx [hunk 7]
```diff
@@ -170,6 +262,24 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
     setSelectedReusableIds([ surveyorId ]);
   };
 
+  /**
+   * Event handler for add reusable confirm
+   */
+  const onImageDialogOpen = (reusable: Reusable) => {
```
Comment (jarinykanen) [inline src/components/survey/reusables.tsx:268]:
```suggestion
  /**
   * Event handler for add reusable confirm
   *
   * @param reusable reusable
   */
  const onImageDialogOpen = (reusable: Reusable) => {
```

File: src/components/survey/reusables.tsx [hunk 8]
```diff
@@ -474,22 +656,209 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+          { renderDeletePreviewButton(index) }
+          <img
+            style={{ maxWidth: "100%", maxHeight: "50vh" }}
+            alt={ uploadedFile.imageUrl }
+            src={ uploadedFile.imageUrl }
+          />
+        </Box>
+      );
+    } if (uploadedFile.file) {
+      return (
```
Comment (jarinykanen) [inline src/components/survey/reusables.tsx:709]:
```suggestion
    }

    if (uploadedFile.file) {
      return (
```

File: src/components/survey/reusables.tsx [hunk 9]
```diff
@@ -474,22 +656,209 @@ const Reusables: React.FC<Props> = ({ surveyId }) => {
...
+          <ThumbnailButton onClick={ () => setDisplayedImageIndex(index) }>
+            <img
+              alt={ uploadedFile.imageUrl }
+              src={ uploadedFile.imageUrl }
+            />
+          </ThumbnailButton>
+        </Grid>
+      );
+    } if (uploadedFile.file) {
+      return (
```
Comment (jarinykanen) [inline src/components/survey/reusables.tsx:749]:
```suggestion
    }

    if (uploadedFile.file) {
      return (
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/components/generic/generic-dialog.tsx (hunk 5 above)
Comment (Chi-Chu319) [reply src/components/generic/generic-dialog.tsx:13]: So the cancel button could be removes when wanted

**Pull request merged**